src/librbd/image/CloneRequest.cc

```diff
@@ -0,0 +1,513 @@
+
+  bool snap_protected;
+  if (*r == 0) {
+    m_p_imctx->snap_lock.get_read();
+    *r = m_p_imctx->is_snap_protected(m_p_imctx->snap_id, &snap_protected);
+    m_p_imctx->snap_lock.put_read();
+  }
+
+  if (*r < 0 || !snap_protected) {
+    m_r_saved = -EINVAL;
+    send_close();
+    return nullptr;
+  }
+
+  send_metadata_list();
+  return nullptr;
+}
+
+template <typename I>
+void CloneRequest<I>::send_metadata_list() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  librados::ObjectReadOperation op;
+  cls_client::metadata_list_start(&op, "", 0);
+
+  using klass = CloneRequest<I>;
+  librados::AioCompletion *comp =
+    create_rados_ack_callback<klass, &klass::handle_metadata_list>(this);
+  m_out_bl.clear();
+  m_p_imctx->md_ctx.aio_operate(m_p_imctx->header_oid,
+				comp, &op, &m_out_bl);
+  comp->release();
+}
+
+template <typename I>
+Context *CloneRequest<I>::handle_metadata_list(int *r) {
+  ldout(m_cct, 20) << this << " " << __func__ << " r=" << *r << dendl;
+
+  if (*r == 0) {
+    bufferlist::iterator it = m_out_bl.begin();
+    *r = cls_client::metadata_list_finish(&it, &m_pairs);
+  }
+
+  if (*r < 0 && *r != -EOPNOTSUPP && *r != -EIO) {
+    lderr(m_cct) << "couldn't list metadata: " << cpp_strerror(*r) << dendl;
+    m_r_saved = *r;
+    send_remove_child();
+  } else if (*r == 0 && !m_pairs.empty()) {
+    send_metadata_set();
+  } else {
+    send_close();
+  }
+
+  return nullptr;
+}
+
+template <typename I>
+void CloneRequest<I>::send_metadata_set() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  librados::ObjectWriteOperation op;
+  cls_client::metadata_set(&op, m_pairs);
+
+  using klass = CloneRequest<I>;
+  librados::AioCompletion *comp =
+    create_rados_ack_callback<klass, &klass::handle_metadata_set>(this);
+  int r = m_ioctx.aio_operate(m_imctx->header_oid, comp, &op);
+  assert(r == 0);
+  comp->release();
+}
+
+template <typename I>
+Context *CloneRequest<I>::handle_metadata_set(int *r) {
+  ldout(m_cct, 20) << this << " " << __func__ << " r=" << *r << dendl;
+
+  if (*r < 0) {
+    lderr(m_cct) << "couldn't set metadata: " << cpp_strerror(*r) << dendl;
+    m_r_saved = *r;
+    send_remove_child();
+  } else {
+    send_close();
+  }
+
+  return nullptr;
+}
+
+template <typename I>
+void CloneRequest<I>::send_close() {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  assert(m_imctx != nullptr);
+
+  using klass = CloneRequest<I>;
+  Context *ctx = create_async_context_callback(
+    *m_imctx, create_context_callback<
+      klass, &klass::handle_close, false>(this));
+  CloseRequest<I> *req = CloseRequest<I>::create(m_imctx, ctx);
+  req->send();
+}
+
+template <typename I>
+Context *CloneRequest<I>::handle_close(int *r) {
+  ldout(m_cct, 20) << this << " " << __func__ << dendl;
+
+  if (*r < 0) {
+    lderr(m_cct) << "couldn't close image: " << cpp_strerror(*r) << dendl;
+    return m_on_finish;
+  }
+
+  if (m_r_saved == 0) {
```

Comment:
Please initialize m_r_saved in the class definition or the constructor. It may be uninitialized at this point.